File: js/examples/openPosition.js
import 'dotenv/config';
import { pathToFileURL } from 'node:url';
import { EZManagerSDK } from '../sdk.js';
import { runReadPosition, printJson } from './readPosition.js';

const POOL_ADDRESS = '0xb2cc224c1c9feE385f8ad6a55b4d94E92359DC59'; // WETH/USDC Aerodrome Example
const OPEN_USDC = '1'; // 1 USDC
const LOWER_PCT = 0.02;
const UPPER_PCT = 0.02;
const SLIPPAGE_PCT = 0.005; // 0.5%

export async function runOpenPosition({
  sdk,
  poolAddress = POOL_ADDRESS,
  usdcAmount = OPEN_USDC,
  lowerPct = LOWER_PCT,
  upperPct = UPPER_PCT,
  slippage = SLIPPAGE_PCT,
  readAfter = true
} = {}) {
  const localSdk = sdk ?? (await EZManagerSDK.fromEnv());

  console.log('Opening position...');
  const result = await localSdk.openPositionByPct({
    poolAddress,
    usdcAmount,
    lowerPct,
    upperPct,
    slippage
  });

  console.log(`Position opened! Tx: ${result.txHash}`);
  if (result.positionKey) console.log(`Key: ${result.positionKey}`);

  const keyToRead = result.positionKey;
  if (!keyToRead) throw new Error('openPositionByPct returned no positionKey; aborting detail lookup.');

  let details = null;
  if (readAfter) {
    details = await runReadPosition({
      sdk: localSdk,
      key: keyToRead,
      label: 'after openPosition',
      blockTag: result?.receipt?.blockNumber ?? 'latest'
    });
  }

  if (!readAfter) {
    printJson({ positionKey: keyToRead });
  }

  return { result, positionKey: keyToRead, details };
}

const isDirectRun = process.argv[1] && import.meta.url === pathToFileURL(process.argv[1]).href;

if (isDirectRun) {
  await runOpenPosition();
}

File: python/examples/openPosition.py
from pathlib import Path
import sys

from dotenv import load_dotenv

sys.path.insert(0, str(Path(__file__).resolve().parents[1]))
from sdk import EZManagerSDK
from readPosition import run_read_position

load_dotenv()

POOL_ADDRESS = '0xb2cc224c1c9feE385f8ad6a55b4d94E92359DC59'  # WETH/USDC Aerodrome Example
OPEN_USDC = '1'  # 1 USDC
LOWER_PCT = 0.02
UPPER_PCT = 0.02
SLIPPAGE_PCT = 0.005  # 0.5%


def run_open_position(
    sdk: EZManagerSDK | None = None,
    pool_address: str = POOL_ADDRESS,
    usdc_amount: str = OPEN_USDC,
    lower_pct: float = LOWER_PCT,
    upper_pct: float = UPPER_PCT,
    slippage: float = SLIPPAGE_PCT,
    read_after: bool = True,
):
    local_sdk = sdk or EZManagerSDK.from_env()

    print('Opening position...')
    result = local_sdk.open_position_by_pct(
        pool_address=pool_address,
        usdc_amount=usdc_amount,
        lower_pct=lower_pct,
        upper_pct=upper_pct,
        slippage=slippage,
    )
    print(f"Position opened! Tx: {result['tx_hash']}")

    position_key = result.get('position_key')
    if position_key:
        print(f'Key: {position_key}')
    else:
        raise RuntimeError('open_position returned no position_key; aborting detail lookup.')

    details = None
    if read_after:
        post_block = (result.get('receipt') or {}).get('blockNumber', 'latest')
        details = run_read_position(
            sdk=local_sdk,
            key=position_key,
            label='after openPosition',
            block_identifier=post_block,
        )

    return {'result': result, 'position_key': position_key, 'details': details}


if __name__ == '__main__':
    run_open_position()